File: app/src/main/java/aromko/de/wishlist/viewModel/WishViewModel.java
package aromko.de.wishlist.viewModel;

import android.app.Application;
import android.net.Uri;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.ViewModel;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import aromko.de.wishlist.database.FirebaseQueryLiveData;
import aromko.de.wishlist.model.Wish;
import aromko.de.wishlist.model.Wishlist;
import aromko.de.wishlist.tasks.AppExecutors;

public class WishViewModel extends ViewModel {
    private static final String DB_PATH_WISHES = "wishes";
    private static final String DB_PATH_WISHLISTS = "wishLists";
    private static DatabaseReference wishes_ref;
    private final FirebaseQueryLiveData liveData;
    private final MediatorLiveData<List<Wish>> listsLiveData = new MediatorLiveData<>();

    public WishViewModel() {
        liveData = new FirebaseQueryLiveData(wishes_ref);
    }

    public WishViewModel(Application mApplication, final String wishlistId) {

        wishes_ref = FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHES + "/" + wishlistId);
        liveData = new FirebaseQueryLiveData(wishes_ref.orderByChild("timestamp"));
        listsLiveData.addSource(liveData, dataSnapshot -> {
            if (dataSnapshot != null) {
                final List<Wish> lists = new ArrayList<>();
                new AppExecutors().mainThread().execute(() -> {
                    for (final DataSnapshot snapshot : dataSnapshot.getChildren()) {
                        Wish wish = snapshot.getValue(Wish.class);
                        wish.setWishId(snapshot.getKey());
                        wish.setWishlistId(wishlistId);
                        lists.add(wish);
                        Collections.sort(lists);
                    }
                    listsLiveData.postValue(lists);
                });

            } else {
                listsLiveData.setValue(null);
            }
        });
    }

    @NonNull
    public LiveData<List<Wish>> getListsLiveData() {
        return listsLiveData;
    }

    public void updateWish(String wishlistId, String wishId, final Wish wish) {
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHES + "/" + wishlistId + "/" + wishId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Wish savedWish = dataSnapshot.getValue(Wish.class);
                wish.setMarkedAsFavorite(savedWish.getMarkedAsFavorite());
                dataSnapshot.getRef().setValue(wish);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    public String insertWish(String wishlistId, Wish wish) {
        String wishId = FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHES).push().getKey();

        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHES + "/" + wishlistId + "/" + wishId).setValue(wish);

        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHLISTS + "/" + wishlistId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Wishlist currentWishlist = dataSnapshot.getValue(Wishlist.class);
                int counter = 1;
                counter += currentWishlist.getWishCounter();
                currentWishlist.setWishCounter(counter);
                dataSnapshot.getRef().setValue(currentWishlist);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
        return wishId;
    }

    public void setWishAsFavorite(String wishlistId, String wishId, Wish wish, String favoriteListId) {
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHES + "/" + wishlistId + "/" + wishId).setValue(wish);
        int counter;
        DatabaseReference databaseReference = FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHES + "/" + favoriteListId + "/" + wishId);
        if (wish.getMarkedAsFavorite() != null && wish.getMarkedAsFavorite().get(FirebaseAuth.getInstance().getCurrentUser().getUid()).equals(true)) {
            databaseReference.setValue(wish);
            counter = 1;
        } else {
            databaseReference.removeValue();
            counter = -1;
        }

        final int finalCounter = counter;
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHLISTS + "/" + favoriteListId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Wishlist wishlist = dataSnapshot.getValue(Wishlist.class);
                wishlist.setWishCounter(wishlist.getWishCounter() + finalCounter);
                if (wishlist.getWishCounter() < 0) {
                    wishlist.setWishCounter(0);
                }
                dataSnapshot.getRef().setValue(wishlist);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    public void selectWish(String wishlistId, String wishId, final FirebaseCallback firebaseCallback) {
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHES + "/" + wishlistId + "/" + wishId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Wish wish = dataSnapshot.getValue(Wish.class);
                firebaseCallback.onCallback(wish);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    public void deleteWish(String wishId, String wishlistId, final String favoriteListId) {
        final int counter = 1;
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHES + "/" + wishlistId + "/" + wishId).removeValue();
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHLISTS + "/" + wishlistId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Wishlist wishlist = dataSnapshot.getValue(Wishlist.class);
                wishlist.setWishCounter(wishlist.getWishCounter() - counter);
                if (wishlist.getWishCounter() < 0) {
                    wishlist.setWishCounter(0);
                }
                dataSnapshot.getRef().setValue(wishlist);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHES + "/" + favoriteListId + "/" + wishId).removeValue();
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHLISTS + "/" + favoriteListId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Wishlist wishlist = dataSnapshot.getValue(Wishlist.class);
                wishlist.setWishCounter(wishlist.getWishCounter() - counter);
                if (wishlist.getWishCounter() < 0) {
                    wishlist.setWishCounter(0);
                }
                dataSnapshot.getRef().setValue(wishlist);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

        deleteImageFromFirebaseStorageByWishId(wishId);
    }

    public void deleteImageFromFirebaseStorageByWishId(String wishId) {
        FirebaseStorage storage = FirebaseStorage.getInstance();
        StorageReference storageRef = storage.getReference();
        StorageReference desertRef = storageRef.child(wishId);
        desertRef.delete().addOnSuccessListener(aVoid -> {
            Log.i(this.getClass().getName(), String.format("File with wishId %s deleted successfully", wishId));
        }).addOnFailureListener(exception -> Log.e(this.getClass().getName(), String.format("File with wishId %s not deleted. More infos: %s", wishId, exception.getMessage())));
    }

    public void updatePhotoUrl(String wishlistId, String wishkey, final Uri uri) {
        FirebaseDatabase.getInstance().getReference("/" + DB_PATH_WISHES + "/" + wishlistId + "/" + wishkey).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                Wish wish = dataSnapshot.getValue(Wish.class);
                wish.setPhotoUrl(uri.toString());
                dataSnapshot.getRef().setValue(wish);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    public interface FirebaseCallback {
        void onCallback(Wish wish);
    }


}

File: app/src/main/java/aromko/de/wishlist/viewModel/WishViewModel.kt
package aromko.de.wishlist.viewModel

import android.app.Application
import android.net.Uri
import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.ViewModel
import aromko.de.wishlist.database.FirebaseQueryLiveData
import aromko.de.wishlist.model.UserSetting
import aromko.de.wishlist.model.Wish
import aromko.de.wishlist.model.Wishlist
import aromko.de.wishlist.tasks.AppExecutors
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.*
import com.google.firebase.storage.FirebaseStorage

class WishViewModel : ViewModel {
    private val liveData: FirebaseQueryLiveData
    val listsLiveData = MediatorLiveData<List<Wish?>?>()

    constructor() {
        liveData = FirebaseQueryLiveData(wishes_ref)
    }

    constructor(mApplication: Application?, wishlistId: String?) {
        wishes_ref = FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES/$wishlistId")
        liveData = FirebaseQueryLiveData(wishes_ref!!.orderByChild("timestamp"))
        listsLiveData.addSource(liveData) { dataSnapshot: DataSnapshot? ->
            if (dataSnapshot != null) {
                val lists:  ArrayList<Wish> = ArrayList()
                AppExecutors().mainThread().execute {
                    for (snapshot in dataSnapshot.children) {
                        val wish = snapshot.getValue(Wish::class.java)
                        wish?.wishId = snapshot.key
                        wish?.wishlistId = wishlistId
                        lists.add(wish!!)
                        lists.sort()
                    }
                    listsLiveData.postValue(lists)
                }
            } else {
                listsLiveData.setValue(null)
            }
        }
    }

    fun getListsLiveData(): LiveData<List<Wish?>?> {
        return listsLiveData
    }

    fun updateWish(wishlistId: String?, wishId: String?, wish: Wish, favoriteListId: String?) {
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES/$wishlistId/$wishId")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    val savedWish = dataSnapshot.getValue(Wish::class.java)
                    wish.markedAsFavorite = savedWish?.markedAsFavorite
                    dataSnapshot.ref.setValue(wish)
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    fun updateFavoriteWishlist(
        favoriteListId: String?,
        wishId: String?,
        wish: Wish
    ) {
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES/$favoriteListId/$wishId")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    val savedWish = dataSnapshot.getValue(Wish::class.java)
                    if (savedWish != null) {
                        wish.markedAsFavorite = savedWish.markedAsFavorite
                        dataSnapshot.ref.setValue(wish)
                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    fun insertWish(wishlistId: String?, wish: Wish?): String? {
        val wishId = FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES").push().key
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES/$wishlistId/$wishId")
            .setValue(wish)
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHLISTS/$wishlistId")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    val currentWishlist = dataSnapshot.getValue(Wishlist::class.java)
                    var counter = 1
                    counter += currentWishlist?.wishCounter!!
                    currentWishlist.wishCounter = counter
                    dataSnapshot.ref.setValue(currentWishlist)
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
        return wishId
    }

    fun setWishAsFavorite(wishlistId: String?, wishId: String?, wish: Wish?, favoriteListId: String?) {
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES/$wishlistId/$wishId").setValue(wish)
        val counter: Int
        val databaseReference = FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES/$favoriteListId/$wishId")
        counter = if (wish?.markedAsFavorite != null && wish.markedAsFavorite!![FirebaseAuth.getInstance().currentUser!!.uid] == true) {
            databaseReference.setValue(wish)
            1
        } else {
            databaseReference.removeValue()
            -1
        }
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHLISTS/$favoriteListId").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val wishlist = dataSnapshot.getValue(Wishlist::class.java)
                wishlist?.wishCounter = wishlist?.wishCounter!! + counter
                if (wishlist.wishCounter < 0) {
                    wishlist.wishCounter = 0
                }
                dataSnapshot.ref.setValue(wishlist)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun selectWish(wishlistId: String?, wishId: String?, firebaseCallback: (Wish) -> Unit) {
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES/$wishlistId/$wishId").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val wish = dataSnapshot.getValue(Wish::class.java)
                firebaseCallback.invoke(wish!!)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun deleteWish(wishId: String?, wishlistId: String?, favoriteListId: String?) {
        val counter = 1
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES/$wishlistId/$wishId").removeValue()
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHLISTS/$wishlistId").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val wishlist = dataSnapshot.getValue(Wishlist::class.java)
                wishlist?.wishCounter = wishlist?.wishCounter!! - counter
                if (wishlist.wishCounter < 0) {
                    wishlist.wishCounter = 0
                }
                dataSnapshot.ref.setValue(wishlist)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES/$favoriteListId/$wishId").removeValue()
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHLISTS/$favoriteListId").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val wishlist = dataSnapshot.getValue(Wishlist::class.java)
                wishlist?.wishCounter = wishlist?.wishCounter!! - counter
                if (wishlist.wishCounter < 0) {
                    wishlist.wishCounter = 0
                }
                dataSnapshot.ref.setValue(wishlist)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
        deleteImageFromFirebaseStorageByWishId(wishId)
    }

    fun deleteImageFromFirebaseStorageByWishId(wishId: String?) {
        val storage = FirebaseStorage.getInstance()
        val storageRef = storage.reference
        val desertRef = storageRef.child(wishId!!)
        desertRef.delete().addOnSuccessListener { aVoid: Void? -> Log.i(this.javaClass.name, String.format("File with wishId %s deleted successfully", wishId)) }.addOnFailureListener { exception: Exception -> Log.e(this.javaClass.name, String.format("File with wishId %s not deleted. More infos: %s", wishId, exception.message)) }
    }

    fun updatePhotoUrl(wishlistId: String, wishkey: String, uri: Uri) {
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_WISHES/$wishlistId/$wishkey").addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                val wish = dataSnapshot.getValue(Wish::class.java)
                wish?.photoUrl = uri.toString()
                dataSnapshot.ref.setValue(wish)
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }

    fun getAllFavoriteWishlistsIdsFromMarkedAsFavoriteUsers(
        userId: String?,
        firebaseCallback: (UserSetting) -> Unit
    ) {
        FirebaseDatabase.getInstance().getReference("/$DB_PATH_SETTINGS/$userId")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    val userSetting = dataSnapshot.getValue(UserSetting::class.java)
                    firebaseCallback.invoke(userSetting!!)
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
    }

    interface FirebaseCallback {
        fun onCallback(wish: Wish?)
    }

    companion object {
        private const val DB_PATH_WISHES = "wishes"
        private const val DB_PATH_WISHLISTS = "wishLists"
        private var wishes_ref: DatabaseReference? = null
        private const val DB_PATH_SETTINGS = "settings"
    }
}